src/archetype/ArchetypeModule.ts

```diff
@@ -3,104 +3,82 @@
-            } else {
-                throw new Error("Unknown generate step.");
-            }
+        const steps: IProjectCreationStep[] = [selectArchetypeStep, specifyArchetypeVersionStep, specifyGroupIdStep, specifyArtifactIdStep, specifyTargetFolderStep];
+        const success: boolean = await runSteps(steps, metadata);
+        if (success) {
+            await executeInTerminalHandler(metadata);
         }
-        return { artifactId: archetypeMetadata.artifactId, groupId: archetypeMetadata.groupId, version: archetypeMetadata.version };
     }
 
-    async function chooseTargetFolder(entry: Uri | undefined): Promise<string> {
-        const result: Uri | undefined = await openDialogForFolder({
-            defaultUri: entry,
-            openLabel: "Select Destination Folder"
-        });
-        const cwd: string | undefined = result !== undefined ? result.fsPath : undefined;
-        if (!cwd) {
-            throw new OperationCanceledError("Target folder not selected.");
-        }
-        return cwd;
+    export async function updateArchetypeCatalog(): Promise<void> {
+        const xml: string = await Utils.downloadFile(REMOTE_ARCHETYPE_CATALOG_URL, true);
+        const archetypes: Archetype[] = await listArchetypeFromXml(xml);
+        const targetFilePath: string = path.join(getPathToExtensionRoot(), "resources", "archetypes.json");
+        await fse.ensureFile(targetFilePath);
+        await fse.writeJSON(targetFilePath, archetypes);
     }
 
-    async function executeInTerminalHandler(archetypeGroupId: string, archetypeArtifactId: string, archetypeVersion: string, targetFolder: string): Promise<void> {
+    async function executeInTerminalHandler(metadata: IProjectCreationMetadata): Promise<void> {
+        const {
+            archetypeArtifactId,
+            archetypeGroupId,
+            archetypeVersion,
+            groupId,
+            artifactId,
+            targetFolder
+        } = metadata;
+        if (archetypeArtifactId === undefined || archetypeGroupId === undefined || archetypeVersion === undefined) {
+            throw new Error("Archetype information is incomplete.");
+        }
+        sendInfo("", { archetypeArtifactId, archetypeGroupId, archetypeVersion });
         const cmdArgs: string[] = [
             // explicitly using 3.1.2 as maven-archetype-plugin:3.0.1 ignores -DoutputDirectory
             // see https://github.com/microsoft/vscode-maven/issues/478
             "org.apache.maven.plugins:maven-archetype-plugin:3.1.2:generate",
             `-DarchetypeArtifactId="${archetypeArtifactId}"`,
             `-DarchetypeGroupId="${archetypeGroupId}"`,
-            `-DarchetypeVersion="${archetypeVersion}"`
+            `-DarchetypeVersion="${archetypeVersion}"`,
+            `-DgroupId="${groupId}"`,
+            `-DartifactId="${artifactId}"`
         ];
         let mvnPath: string | undefined;
-        let cwd: string = targetFolder;
+        let cwd: string | undefined = targetFolder;
```

Comment:
When cwd === undefined, does that mean user press `esc` when specifying the target folder?

Reply:
No. If folder selection is cancelled, nothing will happen, and this method will not be called. Here simply because type definition of `targetFolder` is `string|undefined`. `targetFolder` can not be undefined in real case.